src/plugins/dashboard/public/dashboard_api/get_dashboard_state.test.ts

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import {
+  dataService,
+  embeddableService,
+  savedObjectsTaggingService,
+} from '../services/kibana_services';
+import { getSampleDashboardState } from '../mocks';
+import { DashboardState } from './types';
+import { getDashboardState } from './get_dashboard_state';
+
+dataService.search.searchSource.create = jest.fn().mockResolvedValue({
+  setField: jest.fn(),
+  getSerializedFields: jest.fn().mockReturnValue({}),
+});
+
+dataService.query.timefilter.timefilter.getTime = jest
+  .fn()
+  .mockReturnValue({ from: 'now-15m', to: 'now' });
+
+dataService.query.timefilter.timefilter.getRefreshInterval = jest
+  .fn()
+  .mockReturnValue({ pause: true, value: 0 });
+
+embeddableService.extract = jest
+  .fn()
+  .mockImplementation((attributes) => ({ state: attributes, references: [] }));
+
+if (savedObjectsTaggingService) {
+  savedObjectsTaggingService.getTaggingApi = jest.fn().mockReturnValue({
+    ui: {
+      updateTagsReferences: jest.fn((references, tags) => references),
+    },
+  });
+}
+
+describe('getDashboardState', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should return the current state attributes and references', async () => {
+    const dashboardState = getSampleDashboardState();
+    const result = await getDashboardState({
+      controlGroupReferences: [],
+      generateNewIds: false,
+      dashboardState,
+      panelReferences: [],
+    });
+
+    expect(result.attributes.panels).toEqual([]);
+    expect(result.references).toEqual([]);
+  });
+
+  it('should generate new IDs for panels and references when generateNewIds is true', async () => {
+    const dashboardState = {
+      ...getSampleDashboardState(),
+      panels: { oldPanelId: { type: 'visualization' } },
+    } as unknown as DashboardState;
+    const result = await getDashboardState({
+      controlGroupReferences: [],
+      generateNewIds: true,
+      dashboardState,
+      panelReferences: [
+        {
+          name: 'oldPanelId:indexpattern_foobar',
+          type: 'index-pattern',
+          id: 'bizzbuzz',
+        },
+      ],
+    });
+
+    expect(result.attributes.panels).toEqual(
```

Comment:
Why don't you mock `v4` so you know the new id. Then you can assert attributes.panels has the right shape.

```
jest.mock('uuid', () => ({
  v4: jest.fn().mockReturnValue('12345'),
}));
```

```
expect(result.attributes.panels).toEqual({ ...expectedValue});
```